superduperdb/ext/llm/vllm.py

```diff
@@ -119,51 +121,82 @@ class VllmModel(BaseLLMModel):
 
     def __post_init__(self):
         self.on_ray = self.on_ray or bool(self.ray_address)
+        if 'tensor_parallel_size' not in self.vllm_kwargs:
+            self.vllm_kwargs['tensor_parallel_size'] = self.tensor_parallel_size
+
+        if 'trust_remote_code' not in self.vllm_kwargs:
+            self.vllm_kwargs['trust_remote_code'] = self.trust_remote_code
+
+        if 'model' not in self.vllm_kwargs:
+            self.vllm_kwargs['model'] = self.model_name
+
         super().__post_init__()
 
     def init(self):
-        try:
-            from vllm import LLM
-        except ImportError:
-            raise Exception("You must install vllm with command 'pip install vllm'")
+        class _VLLMCore:
+            """
+            Wrapper for vllm model to support ray.
+            Implementing the client in this way will no longer require vllm dependencies
+            """
+
+            def __init__(self, **kwargs) -> None:
+                try:
+                    from vllm import LLM
+                except ImportError:
+                    raise Exception(
+                        "You must install vllm with command 'pip install vllm'"
+                    )
+                self.model = LLM(**kwargs)
+
+            def generate(self, prompts: List[str], **kwargs) -> List[str]:
+                from vllm import SamplingParams
+
+                sampling_params = SamplingParams(**kwargs)
+                results = self.model.generate(prompts, sampling_params, use_tqdm=False)
+                results = [result.outputs[0].text for result in results]
+                return results
 
         if self.on_ray:
             try:
                 import ray
             except ImportError:
                 raise Exception("You must install vllm with command 'pip install ray'")
 
-            runtime_env = {"pip": ["vllm"]}
+            runtime_env = {
+                "pip": [
+                    "vllm",
+                ]
+            }
             if not ray.is_initialized():
                 ray.init(address=self.ray_address, runtime_env=runtime_env)
 
-            LLM = ray.remote(LLM).remote
+            if self.vllm_kwargs.get('tensor_parallel_size') == 1:
+                # must set num_gpus to 1 to avoid error
```

Comment:
The parallel size here matches the number of GPUs. If it is set to 3, the model will be allocated to 3 GPUs. If the GPU is insufficient, it will wait for GPU resources.

Reply:
Because 1 GPU is used by default, but if it is not specified, the GPU will not be allocated and the GPU will not be found.

Reply:
> may be an assertion during config level so that user knows that num_gpus should be assigned to 1
> 
> else User might be in false hope of 
> 
> num_gpus = 4 and tensor_parallel_size = 1
> 
> 

Will test this case later

Reply:
I’m still a bit unclear about this. Are you suggesting that we should inform users about this behavior? Should this information be communicated through documentation or some specific configuration settings?

Reply:
assert `ray_config.get('num_gpus') == 1` when `self.vllm_kwargs.get('tensor_parallel_size') == 1`, right?

Reply:
yes! something like this

Reply:
once done we can merge this pr :) @jieguangzhou

Reply:
Done, I changed to printing a warning with a description and helping the user set it to 1.
The reason for not using assertions directly is because users will still reset num_gpus, which can increase fault tolerance.
@kartik4949